Redesign: budgets

jorgeatgu opened the pull request:
<!--
NOTE: We're in the middle of a big redesign of the frontend.
That could mean that your PR will not get through on the next few months.
Please see https://github.com/decidim/decidim/discussions/9512
-->

#### :tophat: What? Why?

This PR implements the redesign in budgets module

### :camera: Screenshots

*Please add screenshots of the changes you're proposing*


:hearts: Thank you!

furilo commented:
@decidim/product please review this. As with Debates, take into account pending issues in the description.

carolromero reviewed (changes requested):
@jorgeatgu @furilo sorry for the delay, here's feedback on budgets:
## Mobile
All views are pending.

## Budgets list

- [ ] The "voted budgets" section is not very visible as it is displayed at the end. We need to make it easier to discover. Could you propose an alternative?
- [ ] Hovering on the budgets lists is not working as expected: hovering highlights the card outline but there is no link. Also, if the option is that the whole card is clickable, it is confusing to highlight the title link and the CTA.
https://www.awesomescreenshot.com/video/15863538?key=25941336a8ca523687667350576e11c2
- [ ] Server error when clicking on a budget with no projects
[Enregistrament de pantalla des de 23-3-23 11:31:42.webm](https://user-images.githubusercontent.com/23509895/227176728-7a7c430c-98f6-4ff7-bcf2-9fc0b0f97c08.webm)

## Projects list

- [ ] The budget bar ("shopping cart") should be fixed at the top. 
- [ ] Focus in the project list is not done correctly.

 ![image](https://user-images.githubusercontent.com/23509895/227175749-efe50da9-eaf4-4e2d-9c16-44e3bfa7db40.png)

## Project show page

- [ ] I'm getting a server error when clicking on a project (a couple of days ago it was working, and we had more feedback from the rest of the flow, I'll report it back once I'm able to reproduce it again).

Thanks!

jorgeatgu commented:
> ## Project show page
> * I'm getting a server error when clicking on a project (a couple of days ago it was working, and we had more feedback from the rest of the flow, I'll report it back once I'm able to reproduce it again).
> 
> Thanks!
@carolromero you can continue with the review [staging works fine](https://decidim-redesign.populate.tools/processes/smile-cutting/f/14/budgets/19/projects/50).

carolromero commented:
Hi! Here's the rest of the review:

## Project show page

- [ ] The budget bar ("shopping cart") should be fixed at the top.
- [ ] Remove the "Add to calendar" link, as it is not applicable to Budgets.
- [ ] Place category label as in other Show pages.
![image](https://user-images.githubusercontent.com/23509895/229101225-52c3399c-9646-4fdd-8f00-768054b096f9.png)
- [ ] Not sure if it's being implemented in another PR, but next-prev icons are missing.

### Not logged in 
- [ ] As a visitor, I do not see the pop-up to login or register when trying to add a project to the cart.
- [ ] As a visitor, when clicking on the "Vote budget" button this popup appears.
https://www.awesomescreenshot.com/video/16113086?key=37afb16d100c217ce8136f14595998af

### Logged in 

- [ ] When adding projects to vote, the "Added" tab is not updated.
https://www.awesomescreenshot.com/video/16177697?key=22c101ea42691ae86cd9a2f224920258
- [ ] When clicking on "Back" after checking the details of a project, the warning "You have not yet voted" appears. This popup should only appear when you go from one budget to another (in case there is more than one). Otherwise, as a participant I must be able to enter and exit the project detail page without seeing this warning.
![image](https://user-images.githubusercontent.com/23509895/229465511-92ada917-6a8d-492e-be12-72775b8decbc.png)
- [ ] When confirming the vote, the projects look clickable but are not (see video below). This is in fact the current behavior, but it may make sense to remove the link and make it just an action to confirm or cancel, so as not to complicate the flow further?
- [ ] Confirmation message lacks style (see video).
(https://www.awesomescreenshot.com/video/16180065?key=118760e7ec5ed4c32dc18fb4a2389675)

jorgeatgu commented:
@carolromero Ready to re-review, i have added some comments and videos, not for all points.

> ## Budgets list
> * [x]  The "voted budgets" section is not very visible as it is displayed at the end. We need to make it easier to discover. Could you propose an alternative?

As an alternative, i have placed them at the top.

> * [x]  Hovering on the budgets lists is not working as expected: hovering highlights the card outline but there is no link. Also, if the option is that the whole card is clickable, it is confusing to highlight the title link and the CTA.

![budgets](https://user-images.githubusercontent.com/2649175/231995897-44643c6a-df12-48c4-b7ac-5cb64aeaa899.gif)


> ## Projects list
> * [x]  Focus in the project list is not done correctly.
> 

![projects](https://user-images.githubusercontent.com/2649175/231995876-f09725ae-6bb6-4b5f-9cc2-be67d54d32d5.gif)

> ## Project show page
> * [ ]  Not sure if it's being implemented in another PR, but next-prev icons are missing.

We'll fixed it in another PR. issue: https://github.com/decidim/decidim/issues/10719

> ### Logged in
> * [ ]  When adding projects to vote, the "Added" tab is not updated.

We'll fixed it in another PR, issue: https://github.com/decidim/decidim/issues/10469

> * [x]  When confirming the vote, the projects look clickable but are not (see video below). This is in fact the current behavior, but it may make sense to remove the link and make it just an action to confirm or cancel, so as not to complicate the flow further?

That makes sense. I have removed the link interaction/hover.

carolromero reviewed (changes requested):
@jorgeatgu here's some more feedback:
* Adding a project to the cart is immediately reflected in the progress bar but not in the button (the page has to be refreshed):
https://www.awesomescreenshot.com/video/16685409?key=6dab4af43d5d5b69a23a5df36a40ed6d
* The 'Vote budget' button disabled is clickable and it should not:
https://www.awesomescreenshot.com/video/16685519?key=eef26148fa84f20d57f4264f1d6bc19c

Regarding mobile view, a couple more things:
* Progress bar should be displayed at the top of the project list page.
* In a detail project view, the full version of the progress bar takes up too much space, better to show this more compact version:
![image](https://user-images.githubusercontent.com/23509895/233308245-64606897-8532-40e7-b9ce-6d610b5920ee.png)
* Maybe we should trim the titles of the projects in mobile to avoid this:
![image](https://user-images.githubusercontent.com/23509895/233309127-0273f13f-ba5b-49eb-8c11-4ef0f04c6f3e.png)

Thanks!

carolromero commented:
@jorgeatgu thanks for all the work, the Budgets component it's a tough one :grimacing: I think we a have a last round and we're good to go:

- [ ] When clicking on a related proposal within a project my expectation is to be able to check the content of the proposal without the warning that I am exiting the voting process (check the video below). This is, in fact, how it currently works.
https://www.awesomescreenshot.com/video/17234577?key=8dc17145639009a5c999b7eb923a1932


---
Note: Related to this, we have not quite solved the navigation flow through related content without losing context. In the video you can see that if I go to the related proposal there is a "Back" link. Obviously the expectation of the participant is to go back to where she was, and not go to the proposal index.

This is a problem, not only in Budgets, but in general, since one of the main features of the platform is to be able to link resources of different components to each other to maintain traceability. By the way, this is an ideal use case IMO to implement something like a drawer.

When it comes to navigating backwards we usually have two scenarios:

1. The participant is navigating through related resources (proposals related to a project, to a meeting, etc.) and needs a way to go back to the immediately previous view.
2. The participant lands on the platform via a link to a particular resource (a proposal, a budget project) and needs a way to discover similar ones. 

To solve 2. eventually we will need some other discovery mechanism other than a link that says <- Back. Although we're not going to solve it in this PR I thought I'd leave the feedback anyway (I'm also leaving it noted on the final review board).

furilo commented:
> Obviously the expectation of the participant is to go back to where she was, and not go to the proposal index.

@carolromero totally right.

The idea we entertained when discussing this same problem with the drawers I think its still valid: When linking to a "linked" resource, we may include a param in the URL with the reference to the origin item. 

When viewing the permalink of the linked resource, if the param is present we may show an extra box such as "This item is linked in ORIGIN RESOURCE" with a link to it.

jorgeatgu commented:
> * [ ]  When clicking on a related proposal within a project my expectation is to be able to check the content of the proposal without the warning that I am exiting the voting process (check the video below). This is, in fact, how it currently works.
>   https://www.awesomescreenshot.com/video/17234577?key=8dc17145639009a5c999b7eb923a1932

Done! [propose new features](https://decidim-redesign.populate.tools/processes/roadmap/f/1796/budgets/19/projects/96)

![budget-project-proposals](https://github.com/decidim/decidim/assets/2649175/1f4154fc-1166-4842-b14d-10c0c2104aea)

carolromero reviewed (changes requested):
Thanks @jorgeatgu! Before we move on:

- [ ] There is something going on with the tabs, when switching from one tab to another, the page does not refresh properly.

https://github.com/decidim/decidim/assets/23509895/71deff04-acd7-40ed-95c7-5d161b4cd459

alecslupu commented:
there are some conflicts here ...

jorgeatgu commented:
> there are some conflicts here ...

fixed @alecslupu

alecslupu reviewed (changes requested):
let's enable the RESIGNED_ENABLED variable before the code review, please.

entantoencuanto commented:
> let's enable the RESIGNED_ENABLED variable before the code review, please.

Done

alecslupu commented on decidim-budgets/spec/system/explore_projects_spec.rb:
```diff
@@ -27,6 +27,8 @@
 
     context "when filtering" do
       it "allows searching by text" do
+        skip "REDESIGN_PENDING - Filters integration pending"
```
Is this referring to #10390 ?

alecslupu commented on decidim-budgets/spec/system/explore_projects_spec.rb:
```diff
@@ -70,12 +76,14 @@
       end
 
       it "allows filtering by category" do
+        skip "REDESIGN_PENDING - Filters integration pending"
```
Is this referring to #10390 ?

alecslupu commented on decidim-budgets/spec/system/explore_projects_spec.rb:
```diff
@@ -88,12 +96,14 @@
       end
 
       it "works with 'back to list' link" do
+        skip "REDESIGN_PENDING - Filters integration pending"
```
Is this referring to #10390 ?

alecslupu commented on decidim-budgets/spec/system/explore_projects_spec.rb:
```diff
@@ -129,6 +139,8 @@
         end
 
         it "allows filtering by status" do
+          skip "REDESIGN_PENDING - Filters integration pending"
```
Is this referring to #10390 ?

alecslupu commented on decidim-budgets/spec/system/orders_spec.rb:
```diff
@@ -294,31 +261,37 @@
       end
 
       it "is alerted when trying to leave the component before completing" do
+        skip_unless_redesign_enabled "This test pass with redesigned modals enabled"
```
what is the PR  / Ticket ?

alecslupu reviewed (changes requested):
Partially reviewed. ( 72 / 92 files)

alecslupu commented:
@jorgeatgu are you planning to take this one any time soon ? There are some specs failing.

entantoencuanto commented on decidim-budgets/app/views/decidim/budgets/budgets/index.html.erb:
```diff
@@ -1,5 +1,28 @@
+
+<%= render layout: "layouts/decidim/shared/layout_two_col" do %>
+
+  <% announcement_body = if translated_attribute(current_settings.announcement).present?
```
I'm going to move the announcement_body to a helper method and remove the `if announcement_body.present?` in the call to the cell, because the cell doesn't render nothing if receives a blank argument

ferblape commented:
@alecslupu specs solved

alecslupu commented on decidim-budgets/app/packs/src/decidim/budgets/exit_handler.js:
```diff
@@ -82,7 +84,7 @@ $(() => {
   // not trigger the document link click listener because it has the
   // data-method attribute to trigger a form submit event.
-  $(".header a.sign-out-link").on("click", (event) => {
+  $("[href='/users/sign_out']").on("click", (event) => {
```
Could we add the url to the decidim config variable ?

entantoencuanto replied:
Right!

alecslupu commented on decidim-budgets/app/views/decidim/budgets/projects/_order_progress_progressbar_marks_right.html.erb:
```diff
@@ -0,0 +1,10 @@
+<span class="budget-summary__progressbar-legend budget-summary__progressbar-marks_right" data-total-allocation="<%= current_order.available_allocation %>">
+  <%= t("budget", scope: "decidim.budgets.projects.order_progress") %>
+  <strong class="budget-summary__progressbar-legend-strong">
+    <% if current_order.projects_rule? %>
+    <%= current_order.maximum_projects %>
+    <% else %>
+    <%= budget_to_currency(budget.total_budget) %>
+    <% end %>
```
```suggestion
    <% if current_order.projects_rule? %>
      <%= current_order.maximum_projects %>
    <% else %>
      <%= budget_to_currency(budget.total_budget) %>
    <% end %>
```

alecslupu commented on decidim-budgets/app/views/decidim/budgets/projects/_order_progress_summary_content.html.erb:
```diff
@@ -0,0 +1,115 @@
+        <%= icon "check-line" %>
+      </button>
+    <% end %>
+    <% if include_heading %>
```
we could use a single if statement like: 
```
 <% if include_heading && current_order_checked_out? %>
```

alecslupu commented on decidim-budgets/app/views/decidim/budgets/projects/_order_progress_summary_content.html.erb:
```diff
@@ -0,0 +1,115 @@
+            <%= icon "check-line" %>
+          </button>
+        <% end %>
+        <% if include_heading %>
```
<% if include_heading && current_order_checked_out? %>

alecslupu commented on decidim-budgets/app/views/decidim/budgets/projects/_order_total_budget.html.erb:
```diff
@@ -1,7 +1,8 @@
-<strong id="order-total-budget" class="mini-title__strong">
+<strong id="order-total-budget" class="budget-summary__progressbar-legend-strong">
   <% if current_order.projects_rule? %>
```
indentation

alecslupu commented on decidim-budgets/app/views/decidim/budgets/projects/_order_progress_summary_content.html.erb:
```diff
@@ -0,0 +1,111 @@
+<div <%= responsive ? "data-order-progress-responsive=true" : "data-order-progress-responsive=false" %> class="budget-summary__content">
+  <div class="budget-summary__progressbox" data-current-allocation="<%= current_order ? current_order.total : 0 %>">
+    <% if include_heading %>
+      <% if current_order_checked_out? %>
+        <%= t("title", scope: "decidim.budgets.projects.budget_summary.checked_out") %>
+        <div class="budget-summary__content__header">
+          <span class="budget-summary__content__header--title">
+            <%= t("name", scope: "decidim.components.budgets") %>
+          </span>
+          /
+          <%= translated_attribute(budget.title) %>
+          <p class="budget-summary__content__header--description">
+            <%= t("description", scope: "decidim.budgets.projects.budget_summary.checked_out") %>
+          </p>
+        </div>
+      <% end %>
+    <% end %>
+    <% if include_heading && !current_order_checked_out? %>
+      <p class="budget-summary__progressbox--checked_out">
+        <%= current_rule_description %>
+      </p>
+    <% end %>
```
Sorry, i have seen it after the last review session... 
```suggestion
    <% if include_heading %>
      <% if current_order_checked_out? %>
        <%= t("title", scope: "decidim.budgets.projects.budget_summary.checked_out") %>
        <div class="budget-summary__content__header">
          <span class="budget-summary__content__header--title">
            <%= t("name", scope: "decidim.components.budgets") %>
          </span>
          /
          <%= translated_attribute(budget.title) %>
          <p class="budget-summary__content__header--description">
            <%= t("description", scope: "decidim.budgets.projects.budget_summary.checked_out") %>
          </p>
        </div>
      <% else %>
        <p class="budget-summary__progressbox--checked_out">
          <%= current_rule_description %>
        </p>
      <% end %>
    <% end %>
```

alecslupu reviewed (changes requested):
One small change, i have seen after i have provided last feedback session. And this can be merged.

ferblape commented:
You are right about the change
